### ISSUE TITLE
Unhandled exception: Cast from Nil to Hash(K, V) failed

### ISSUE THREAD

**Issue by exhaustedMutex** [hahwul/noir#126]
Hi I can not solve this issue, tried to update/reinstall both noir and crystal but none of them helped.

Any idea how can I solve this?

`    Analysis to 2 technologies
Unhandled exception: Cast from Nil to Hash(K, V) failed, at /home/linuxbrew/.linuxbrew/Cellar/crystal/1.9.2/share/crystal/src/yaml/any.cr:286:5:286 (TypeCastError)
  from noir in '??'
  from noir in '??'
  from noir in '??'
  from noir in '__crystal_main'
  from noir in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from noir in '_start'
  from ???`

**Comment by hahwul** [comment]
Hi
First, thank you for submit this issue. I think.. It looks like a bug in noir. I'll check it out :)

**Comment by exhaustedMutex** [comment]
> Hi First, thank you for submit this issue. I think.. It looks like a bug in noir. I'll check it out :)

No worries, Thanks for the quick response.

**Comment by exhaustedMutex** [comment]
@hahwul 

Hi thanks for following this.

here is the output:

```
 › ./bin/noir -b .
              .__
  ____   ____ |__|______
 /    \ /  _ \|  \_  __ \
|   |  (  <♠> )  ||  | \/
|___|  /\____/|__||__|
     \/                 v0.9.1

[*] Detecting technologies to base directory.
[I] Detected 2 technologies.
    oas3
    js_express
[*] Initiate code analysis based on the detected technology.
[*] Starting analysis of endpoints.
    18 Analyzers initialized
    Analysis to 2 technologies
Unhandled exception: Cast from Nil to Hash(K, V) failed, at /usr/lib/crystal/lib/yaml/any.cr:286:5:286 (TypeCastError)
  from /usr/lib/crystal/lib/yaml/any.cr:285:3 in 'as_h'
  from src/analyzer/analyzers/analyzer_oas3.cr:102:11 in 'analyze'
  from src/analyzer/analyzers/analyzer_oas3.cr:160:3 in 'analyzer_oas3'
  from src/analyzer/analyzer.cr:19:23 in '->'
  from src/analyzer/analyzer.cr:49:40 in 'analysis_endpoints'
  from src/models/noir.cr:71:18 in 'analyze'
  from src/noir.cr:118:3 in '__crystal_main'
  from /usr/lib/crystal/lib/crystal/main.cr:129:5 in 'main_user_code'
  from /usr/lib/crystal/lib/crystal/main.cr:115:7 in 'main'
  from /usr/lib/crystal/lib/crystal/main.cr:141:3 in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from ./bin/noir in '_start'
  from ???
```

To avoid confusion and wasting your time I'm gonna explain how I use the tool.
I am working on a bug bounty program that source is provided for self-hosting the service.
I cloned the repo and ran `noir -b .` in the directory. I also tried to create a free account on the service website and tried noir with `noir -b . -u https://mysub.service.com` which again gives me the same result.

**Comment by hahwul** [comment]
@exhaustedMutex 
Wow, thank you so much! It's going to be a great help in identifying the issue. 
I'll try to make the fix as soon as possible and push it to the dev branch :D

**Comment by exhaustedMutex** [comment]
@hahwul No worries. Happy to help. Looking forward to try the fix.

**Comment by hahwul** [comment]
@exhaustedMutex 
Hi, I added the missing exception handling in the code. 
Would it still reproduce in the current dev branch version?

```bash
git pull -v 
shards build
```

**Comment by exhaustedMutex** [comment]
@hahwul  Nice one. Thank you so much. It does work now :)

but for some reason even when I provide -u url it doesn't provide the full url. Also none of POST methods have any data. When I check the source code there are some API endpoints which are missing in the report.

Would you be interested to provide you the target program link to see have a look yourself?

```
[*] Detecting technologies to base directory.
[I] Detected 2 technologies.
    oas3
    js_express
[*] Initiate code analysis based on the detected technology.
[*] Starting analysis of endpoints.
    18 Analyzers initialized
    Analysis to 2 technologies
    15 endpoints found
[*] Optimizing endpoints.
[I] Finally identified 15 endpoints.
[*] Generating Report.
GET /  
POST /setup
```

**Comment by hahwul** [comment]
Hi @exhaustedMutex 
In the case of js express, its [current support range is quite limited](https://github.com/hahwul/noir#languages-and-frameworks) (URL + Method). This means that it's likely to have less information compared to other analyzers. The -u option also doesn't seem to work, and this issue appears to be due to the fact that the js express code has been largely neglected. 

So, I'll modifications to enable the -u option to work globally in this patch, and gradually work on expanding the functionality of js express in the future.

**Comment by hahwul** [comment]
## Test
![](https://github.com/hahwul/noir/assets/13212227/6a1900c2-f848-4569-98d6-27a282841b74)

**Comment by hahwul** [comment]
@exhaustedMutex 
Hello again :D
The bug you reported appears to have been fixed, and closing the issue. It has been addressed in [v0.9.1](https://github.com/hahwul/noir/releases/tag/v0.9.1), and you can upgrade using the following method. If it happens to reoccur, please feel free to reopen the issue :D

```bash
# If you use home brew
brew upgrade

# If you do not see the updated version, please run this command.
brew update; brew upgrade
```

```bash
# Install with source
git clone https://github.com/hahwul/noir; cd noir
shards install
shards build
```

**Comment by exhaustedMutex** [comment]
@hahwul Nice one. It works without any issue now. Thanks for the great tool xx
